File: packages/vscode-extension/src/evolution-deployment-config.js
"use strict";

const { runCliResult } = require("./chat/introspect-commands.js");

// Render the CLI's admission projection; this never infers runtime authority.
function deploymentReadinessRows(status) {
  return ["ask", "agent"].map((command) => {
    const value = status?.readiness?.[command];
    const admitted = value?.state === "admitted";
    const known =
      value?.scope === "deployment-admission" &&
      value.runtimeVerification === "not_checked" &&
      Array.isArray(value.requiredCommands) &&
      value.requiredCommands.includes(command) &&
      (admitted
        ? value.ready === true && value.taskReady === null
        : [
            "not_configured",
            "disabled",
            "invalid",
            "command_not_allowed",
          ].includes(value.state) &&
          value.ready === false &&
          value.taskReady === false);
    return {
      command,
      state: known ? value.state : "unknown",
      summary: known
        ? admitted
          ? "部署准入通过"
          : "部署准入被阻断"
        : "未提供准入诊断",
      detail:
        known && typeof value.detail === "string"
          ? value.detail
          : "当前 CLI 未提供可识别的部署准入诊断，请更新 CLI。",
      remediation:
        known && typeof value.remediation === "string" ? value.remediation : "",
    };
  });
}

function parseStatus(text) {
  let value;
  try {
    value = JSON.parse(String(text || ""));
  } catch {
    throw new Error("cc 返回了无法识别的 Skill 自进化配置结果");
  }
  if (!value || typeof value !== "object" || Array.isArray(value))
    throw new Error("cc 返回了无效的 Skill 自进化配置结果");
  if (value.ok === false) throw new Error(value.error || "配置操作失败");
  return value;
}

async function run(command, args, { cwd, deps } = {}) {
  const result = await runCliResult({
    command,
    args: ["evolution", "deployment", ...args, "--json"],
    cwd,
    timeoutMs: 60_000,
    deps,
  });
  if (!result.ok)
    throw new Error(result.text || result.error || "无法运行 cc 配置命令");
  return parseStatus(result.stdout || result.text);
}

function getEvolutionDeploymentStatus(options = {}) {
  return run(options.command, ["status"], options);
}

function configureEvolutionDeployment(
  { descriptorPath, trustRootPath },
  options = {},
) {
  if (!descriptorPath || !trustRootPath)
    throw new Error("请选择签名描述符和 Ed25519 信任根公钥");
  return run(
    options.command,
    [
      "configure",
      "--descriptor",
      descriptorPath,
      "--trust-root",
      trustRootPath,
    ],
    options,
  );
}

function setEvolutionDeploymentEnabled(enabled, options = {}) {
  return run(options.command, [enabled ? "enable" : "disable"], options);
}

module.exports = {
  deploymentReadinessRows,
  parseStatus,
  getEvolutionDeploymentStatus,
  configureEvolutionDeployment,
  setEvolutionDeploymentEnabled,
};

File: packages/jetbrains-plugin/src/main/java/com/chainlesschain/ide/EvolutionDeploymentConfig.java
package com.chainlesschain.ide;

import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/** Fixed argv and fail-closed parsing for the shared cc evolution deployment profile. */
public final class EvolutionDeploymentConfig {
    private EvolutionDeploymentConfig() {}

    /** Display-only admission; no runtime authority is inferred from this value. */
    public record Admission(String state, String detail, String remediation) {
        public boolean known() { return !"unknown".equals(state); }
        public boolean admitted() { return "admitted".equals(state); }
    }

    public record Status(boolean effectiveEnabled, boolean profileEnabled,
                         boolean verified, String source, String descriptorPath,
                         String trustRootPath, String profilePath, String error,
                         List<String> commands, Map<String, Admission> readiness) {}

    public static List<String> statusArgs() {
        return List.of("evolution", "deployment", "status", "--json");
    }

    public static List<String> configureArgs(String descriptorPath, String trustRootPath) {
        if (descriptorPath == null || descriptorPath.isBlank()
                || trustRootPath == null || trustRootPath.isBlank()) {
            throw new IllegalArgumentException("Descriptor and trust-root paths are required");
        }
        return List.of("evolution", "deployment", "configure", "--descriptor",
                descriptorPath.trim(), "--trust-root", trustRootPath.trim(), "--json");
    }

    public static List<String> toggleArgs(boolean enabled) {
        return List.of("evolution", "deployment", enabled ? "enable" : "disable", "--json");
    }

    public static Status parseStatus(String json) {
        Map<String, Object> value = MiniJson.parseObject(json == null ? "" : json.trim());
        if (Boolean.FALSE.equals(value.get("ok"))) {
            throw new IllegalArgumentException(text(value.get("error")));
        }
        List<String> commands = new ArrayList<>();
        Object rawCommands = value.get("commands");
        if (rawCommands instanceof List<?> list) {
            for (Object command : list) if (command instanceof String s) commands.add(s);
        }
        return new Status(Boolean.TRUE.equals(value.get("effectiveEnabled")),
                Boolean.TRUE.equals(value.get("profileEnabled")),
                Boolean.TRUE.equals(value.get("verified")), text(value.get("source")),
                text(value.get("descriptorPath")), text(value.get("trustRootPath")),
                text(value.get("profilePath")), text(value.get("error")), List.copyOf(commands),
                Map.of("ask", parseAdmission(value.get("readiness"), "ask"),
                        "agent", parseAdmission(value.get("readiness"), "agent")));
    }

    private static Admission parseAdmission(Object raw, String command) {
        Admission unknown = new Admission("unknown", null, null);
        if (!(raw instanceof Map<?, ?> entries) || !(entries.get(command) instanceof Map<?, ?> value))
            return unknown;
        String state = text(value.get("state"));
        boolean admitted = "admitted".equals(state);
        boolean known = "deployment-admission".equals(value.get("scope"))
                && "not_checked".equals(value.get("runtimeVerification"))
                && value.get("requiredCommands") instanceof List<?> required && required.contains(command)
                && (admitted
                    ? Boolean.TRUE.equals(value.get("ready")) && value.containsKey("taskReady") && value.get("taskReady") == null
                    : state != null && List.of("not_configured", "disabled", "invalid", "command_not_allowed").contains(state)
                        && Boolean.FALSE.equals(value.get("ready")) && Boolean.FALSE.equals(value.get("taskReady")));
        return known ? new Admission(state, text(value.get("detail")), text(value.get("remediation"))) : unknown;
    }

    public static Status run(List<String> args, File cwd, long timeoutMs) {
        String output = AgentChatSession.runCapture(args, cwd, timeoutMs);
        if (output == null || output.isBlank()) {
            throw new IllegalStateException("cc did not return a configuration result; update the CLI and retry");
        }
        return parseStatus(output);
    }

    private static String text(Object value) {
        return value instanceof String s && !s.isBlank() ? s : null;
    }
}
